third_party/gfxreconstruct/framework/decode/vulkan_dive_consumer_base.cpp

```diff
@@ -0,0 +1,185 @@
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#include "decode/vulkan_dive_consumer_base.h"
+#include "decode/custom_vulkan_struct_to_json.h"
+
+#include "generated/generated_vulkan_enum_to_json.h"
+
+#include "util/json_util.h"
+#include "util/platform.h"
+#include "util/file_path.h"
+#include "util/to_string.h"
+
+GFXRECON_BEGIN_NAMESPACE(gfxrecon)
+GFXRECON_BEGIN_NAMESPACE(decode)
+
+using namespace util::platform;
+
+VulkanExportDiveConsumerBase::VulkanExportDiveConsumerBase() {}
+
+VulkanExportDiveConsumerBase::~VulkanExportDiveConsumerBase()
+{
+    Destroy();
+}
+
+void VulkanExportDiveConsumerBase::Initialize(AnnotationHandler* writer)
+{
+    GFXRECON_ASSERT(writer);
+    writer_ = writer;
+}
+
+void VulkanExportDiveConsumerBase::Destroy()
+{
+}
+
+void VulkanExportDiveConsumerBase::ProcessSetDeviceMemoryPropertiesCommand(
+    format::HandleId                             physical_device_id,
+    const std::vector<format::DeviceMemoryType>& memory_types,
+    const std::vector<format::DeviceMemoryHeap>& memory_heaps)
+{
+}
+
+void VulkanExportDiveConsumerBase::Process_vkCmdBuildAccelerationStructuresIndirectKHR(
+    const ApiCallInfo&                                                         call_info,
+    format::HandleId                                                           commandBuffer,
+    uint32_t                                                                   infoCount,
+    StructPointerDecoder<Decoded_VkAccelerationStructureBuildGeometryInfoKHR>* pInfos,
+    PointerDecoder<VkDeviceAddress>*                                           pIndirectDeviceAddresses,
+    PointerDecoder<uint32_t>*                                                  pIndirectStrides,
+    PointerDecoder<uint32_t*>*                                                 ppMaxPrimitiveCounts)
+{
+    nlohmann::ordered_json dive_data;
+    const util::JsonOptions json_options;
+    auto& args = dive_data["args"];
+    HandleToJson(args["commandBuffer"], commandBuffer, json_options);
+    FieldToJson(args["infoCount"], infoCount, json_options);
+    FieldToJson(args["pInfos"], pInfos, json_options);
+    FieldToJson(args["pIndirectDeviceAddresses"], pIndirectDeviceAddresses, json_options);
+    FieldToJson(args["pIndirectStrides"], pIndirectStrides, json_options);
+
+    auto infos                     = pInfos ? pInfos->GetPointer() : nullptr;
+    auto max_primitive_counts      = ppMaxPrimitiveCounts ? ppMaxPrimitiveCounts->GetPointer() : nullptr;
+    auto max_primitive_counts_json = args["ppMaxPrimitiveCounts"];
+
+    for (uint32_t i = 0; i < infoCount; ++i)
+    {
+        auto element = max_primitive_counts_json[i];
```

Comment:
element is created inside the loop as `auto element = max_primitive_counts_json[i];`. And this creates a copy, and then FieldToJson operates on the copy. is this intentional?